code/krepel/container/dynamic_array.d

```diff
@@ -0,0 +1,148 @@
+module krepel.container.dynamic_array;
+
+import krepel.memory;
+import krepel.algorithm : Max;
+import Meta = krepel.meta;
+
+struct DynamicArray(T, A = typeof(null))
+{
+  enum size_t MinimumElementAllocationCount = 16;
+
+  alias ElementType = T;
+
+  static if(is(A == typeof(null)))
+  {
+    alias AllocatorType = typeof(GlobalAllocator);
+    AllocatorType* AllocatorPtr = &GlobalAllocator;
+  }
+  else
+  {
+    alias AllocatorType = A;
+    AllocatorType* AllocatorPtr;
+  }
+
+  ElementType[] Capacity;
+
+  // Is always a subset of Capacity.
+  ElementType[] Data;
+
+  @property auto ref Allocator() inout { return *AllocatorPtr; }
+
+  this(ref AllocatorType Allocator)
+  {
+    AllocatorPtr = &Allocator;
+  }
+
+  inout(ElementType)[] opSlice(size_t LeftIndex, size_t RightIndex) inout
+  {
+    return Data[LeftIndex .. RightIndex];
+  }
+
+  inout(ElementType)[] opIndex() inout { return Data; }
+
+  ref inout(ElementType) opIndex(IndexType)(IndexType Index) inout
+   if(Meta.IsIntegral!IndexType)
+  {
+    return Data[Index];
+  }
+
+  inout(ElementType)[] opIndex(T)(T[] Slice) inout
```

Comment:
It's the counter-part to `opSlice` which enables you to write something like this:

```d
auto SomeSlice = MyArray[0..2]; // Get the first two elements.
// Will roughly translate to:
// auto SomeSlice = MyArray.opIndex(MyArray.opSlice(0, 2));
```